### PR TITLE
Return a customized error message when there is an error during function verification or execution.

### PR DESCRIPTION
**Pull request by steelgeek091**
When there is an error in MoveVM verification or execution function, return a customized error message which includes the module and function name.

resolve https://github.com/rooch-network/rooch/issues/227.

### CODE REVIEW

File: crates/rooch-types/src/transaction/mod.rs [hunk 1]
```diff
@@ -170,7 +171,7 @@ pub struct TransactionInfo {
     /// The vm status. If it is not `Executed`, this will provide the general error class. Execution
     /// failures and Move abort's receive more detailed information. But other errors are generally
     /// categorized with no status code or other information
-    pub status: KeptVMStatus,
+    pub status: VmStatusExplainView,
```
Comment (jolestar) [inline crates/rooch-types/src/transaction/mod.rs:174]: the View type only for the JSON RPC output, the transaction info should keep the original Status

File: moveos/moveos/src/vm/vm_status_explainer.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        location: AbortLocation,
+        function: u16,
+        // The function name refers to the function that fails during MoveVM verify or execution,
+        // not the function called by the user.
+        function_name: Option<String>,
+        code_offset: u16,
+    },
+}
+
+pub fn explain_vm_status(state_db: &StateDB, vm_status: VMStatus) -> Result<VmStatusExplainView> {
```
Comment (jolestar) [inline moveos/moveos/src/vm/vm_status_explainer.rs:52]: 这里的 StateDB 可以用 `ModuleResolver`

File: crates/rooch-types/src/transaction/mod.rs (hunk 1 above)
Comment (steelgeek091) [reply crates/rooch-types/src/transaction/mod.rs]: Okay, it has been fixed.

File: moveos/moveos/src/vm/vm_status_explainer.rs (hunk 2 above)
Comment (steelgeek091) [reply moveos/moveos/src/vm/vm_status_explainer.rs:52]: I have tried, but due to borrowing issues with the Session struct, I need to change the code structure in order to solve it.

Here is the function being called:

```rust
let (change_set, raw_events, mut extensions) = session.finish_with_extensions()?;`
```

`session.finish_with_extension(self)` is a move type, which means that session cannot be used afterwards.

```rust
pub fn finish_with_extensions(
        self,
    ) -> VMResult<(ChangeSet, Vec<Event>, NativeContextExtensions<'r>)> {
        self.session.finish_with_extensions()
    }
```

File: moveos/moveos/src/moveos.rs [hunk 3]
```diff
@@ -265,9 +271,13 @@ impl MoveOS {
                         .finish(Location::Undefined)
                 })?;
 
+                let state_db = self.db.get_state_store();
+                let vm_status_explain = explain_vm_status(state_db, vm_status_cloned)?;
+
                 Ok(TransactionOutput {
                     state_root: new_state_root,
                     status,
+                    status_reexplain: vm_status_explain,
```
Comment (baichuan3) [inline moveos/moveos/src/moveos.rs:280]: vm_status_explain contains status ?

File: crates/rooch-types/src/transaction/mod.rs [hunk 4]
```diff
@@ -171,6 +173,8 @@ pub struct TransactionInfo {
     /// failures and Move abort's receive more detailed information. But other errors are generally
     /// categorized with no status code or other information
     pub status: KeptVMStatus,
+
+    pub status_reexplain: VmStatusExplainView,
```
[the remark this answers is not in the document]
Comment (steelgeek091) [reply crates/rooch-types/src/transaction/mod.rs]: Okay, the field has been removed.

File: moveos/moveos/src/moveos.rs [hunk 5]
```diff
@@ -39,6 +42,7 @@ pub struct TransactionOutput {
     /// The new state root after the transaction execution.
     pub state_root: H256,
     pub status: KeptVMStatus,
+    pub status_reexplain: VmStatusExplainView,
```
Comment (jolestar) [inline moveos/moveos/src/moveos.rs:45]: VmStatusExplainView should not be part of TransactionOutput. But we can refactor in #239

**Review by jolestar — APPROVED** [review]

**Pull request merged**